### ISSUE TITLE
Some blackbox tests fail (on macos)

### ISSUE THREAD

**Issue by axxel** [nu-book/zxing-cpp#34]
Working on the recent DMDetector bug, I was surprised to note that there are few unrelated blackbox tests that fail (see blow). Looking for a regression, I went back until I reached pretty much the first commit that allowed me to execute the tests on macos and all the tested versions fail for those files?!?

Trying to make sure there are no compiler dependent #ifdefs or undefined behaviors causing the problem, I came up with #33.

Still no idea what is going on with those files on my system. Question @huycn: Do all blackbox tests pass with your windows builds?

Here is a list of failing tests:
```
code128-2 @ 180°, total:  40, fast:  35 of  36
  FAILED: Not detected (fast): 11.png 12.png 28.png 30.png 37.png
ean13-1      @   0°, total:  34, fast:  29 of  30
  FAILED: Not detected (fast): 20.png 34.png 35.png 36.png 37.png
upca-1        @   0°, total:  21, slow:  17 of  18
  FAILED: Not detected (slow): 1.png 18.png 20.png
upca-1        @ 180°, total:  21, fast:  13 of  16
  FAILED: Not detected (fast): 1.png 10.png 12.png 18.png 19.png 20.png 3.png 4.png
     , slow:  17 of  18
  FAILED: Not detected (slow): 1.png 18.png 20.png
```

plus a bunch of qrcode tests.

**Comment by huycn** [comment]
On my Windows, all tests are passed as expected.
When I tried on Linux: I got other results which is not like those on my Windows, neither like yours:
Only qrcode-4 didn't pass:
```
FAILED: Not detected (fast): 05.png 06.png 08.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png
, slow:  34 of  36, misread: 0 of 0
FAILED: Not detected (slow): 05.png 06.png 08.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png

qrcode-4             @  90°, total:  48, fast:  33 of  35, misread: 0 of 0
FAILED: Not detected (fast): 04.png 05.png 06.png 08.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png
, slow:  33 of  35, misread: 0 of 0
FAILED: Not detected (slow): 04.png 05.png 06.png 08.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png

qrcode-4             @ 180°, total:  48, fast:  33 of  35, misread: 0 of 0
FAILED: Not detected (fast): 05.png 06.png 08.png 09.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png
, slow:  33 of  35, misread: 0 of 0
FAILED: Not detected (slow): 05.png 06.png 08.png 09.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png

qrcode-4             @ 270°, total:  48, fast:  33 of  35, misread: 0 of 0
FAILED: Not detected (fast): 05.png 06.png 08.png 09.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png
, slow:  33 of  35, misread: 0 of 0
FAILED: Not detected (slow): 05.png 06.png 08.png 09.png 13.png 14.png 19.png 24.png 28.png 30.png 31.png 32.png 33.png 34.png 41.png
```
That is intriguing, I will compare to see what is going on.

**Comment by axxel** [comment]
I suddenly remembered https://github.com/nu-book/zxing-cpp/commit/a66a8dbdca61af41f659686e5097220b73d7b494#diff-18ce59a849129be1194439faabba4c44. So the 2 failing qrcode-4 ones you see are possibly still related to some difference between the rgb->grey conversion. Any maybe my 'regression' with the above mentioned 1D ones is the same problem and the difference between your Linux/my old macos results and my current macos results is simply an updated imagemagick on my macos system?

This 'brittleness' of that simple conversion is really unfortunate (and wired). If you could tell me the exact formula that your GDI+ path is using for the rgb->grey conversion, I could use that instead of relying on imagemagick. Hm, IIRC I checked that the last time and told imagemagick exactly what to do. That's still the best idea I have right now.

**Comment by huycn** [comment]
For GDI+, the RGB->grey conversion is done by GenericLuminanceSource by using its other constructor. I have tried to change the generic runner as following:
```
diff --git a/test/runners/generic/TestReaderMain.cpp b/test/runners/generic/TestReaderMain.cpp
index 58c71af..3919356 100644
--- a/test/runners/generic/TestReaderMain.cpp
+++ b/test/runners/generic/TestReaderMain.cpp
@@ -52,19 +52,19 @@ static std::vector<fs::path> getImagesInDirectory(const fs::path& dirPath)
 static std::shared_ptr<LuminanceSource> readPNM(FILE* f)
 {
 	int w, h;
-	if (fscanf(f, "P5\n%d %d\n255\n", &w, &h) != 2)
+	if (fscanf(f, "P6\n%d %d\n255\n", &w, &h) != 2)
 		throw std::runtime_error("Failed to parse PNM file header.");
-	auto ba = std::make_shared<ByteArray>(w * h);
-	auto read = fread(ba->data(), sizeof(uint8_t), w*h, f);
+	auto ba = std::make_shared<ByteArray>(3 * w * h);
+	auto read = fread(ba->data(), sizeof(uint8_t), 3*w*h, f);
 //	if (read != w * h)
 //		throw std::runtime_error("Failed to read PNM file data: " + std::to_string(read) + " != " + std::to_string(w * h) + " -> " + std::to_string(feof(f)));
-	return std::make_shared<GenericLuminanceSource>(0, 0, w, h, ba, w);
+	return std::make_shared<GenericLuminanceSource>(w, h, ba->data(), 3*w, 3, 0, 1, 2);
 }
 
 static std::shared_ptr<LuminanceSource> readImage(const fs::path& filename)
 {
-	std::string cmd = "convert '" + filename.native() + "' -intensity Rec601Luma -colorspace gray +set comment pgm:-";
-	bool pipe = filename.extension() != ".pgm";
+	std::string cmd = "convert '" + filename.native() + "' +set comment ppm:-";
+	bool pipe = filename.extension() != ".ppm";
 	FILE* f = pipe ? popen(cmd.c_str(), "r") : fopen(filename.c_str(), "r");
 	if (!f)
 		throw std::runtime_error("Failed to open pipe '" + cmd + "': " + std::strerror(errno));
```
But the results are worse, many more tests didn't pass. I will definitively look further into it to see where these differences come from.

**Comment by huycn** [comment]
I've compared qrcode-4/24.png and qrcode-4/28.png which are two that pass on my Windows but don't on my Linux and with my change above, the RGB inputs to `GenericLuminanceSource` are exactly the same on both platform. That suggests that the real problem is laying somewhere else.

**Comment by huycn** [comment]
The problem was that the grayscale image produced by ImageMagic is not the same as using the simple conversion method `RGBToGray()` in `GenericLuminanceSource.cpp`. I've tried to convert the PNG to PPM color format instead of PGM but for a reason I ignore, it fixed some tests but failed other tests. Thus, I integrated LodePNG (from https://github.com/lvandeve/lodepng) to read directly .png files. It gives now the same results as on GDI+ runner. Also, it runs faster than before, obviously.
